### ISSUE TITLE
Properly define how to merge Pull Requests

### ISSUE THREAD

**Issue by NicoHambauer** [amosproj/amos2023ws06-sales-lead-qualifier#25]
Standardised strategy to merge Pull Requests was asked for.
Make sure this is also used regularly.
So this strategy should be simple.

Definitely do this, as this was addressed 3 times in the retro 1

**Comment by Tims777** [comment]
Maybe somebody else than Felix can do this, as he is already tasked with one of our largest items for this sprint.

**Comment by Tims777** [comment]
My idea is that somebody can read on / think of different ways to handle this and maybe create a poll in the end so that everybody can have a say.

**Comment by ultiwinter** [comment]
**main** and **dev** branches are protected against direct pushes, which means that we want to do a Pull Request (PR) in order to merge a developed branch into these branches. Having developed a branch (let's call it **feature-1**) and we want to merge **feature-1** branch into **main** branch.

Here is a standard way to merge pull requests:

1. Have all your local changes added, committed, and pushed on the remote **feature-1** branch
> git checkout feature-1
> git add .
> git commit -m "added a feature" --signoff  (don't forget the signoff ;) )
>git push

2. Make sure your local main branch up-to-date 
> git checkout main
> git pull main

3. Go to Pull Requests (https://github.com/amosproj/amos2023ws06-sales-lead-qualifier/pulls) > click on "New pull request" > make sure the base is **main** branch (or **dev** branch, depends on which branch you want to update) and the compare to be your **feature-1** branch, as highlighted in the photo below and click "create pull requests":
![image](https://github.com/amosproj/amos2023ws06-sales-lead-qualifier/assets/95130956/3eceb19d-bdb7-41bd-aa58-07ed7fb6148a)

4. wait for others to review your code, when we agree to the changes we would merge the PR.

_**In case of merge conflict:**_

Should we experience merge conflict after step 3, we should solve the merge conflicts manually, below the title of "This branch has conflicts that must be resolved" click on web editor (you can use vscode or any editor you want).
The conflict should look like this:
>"<<<<<<< HEAD"
>"// Your changes at **feature-1** branch"
>"======="
>"// Data already on the main branch"
>">>>>>>> main"

-choose which one of these you would adopt for the merge to the **main** branch, we would be better off solving the merge -conflicts together rather than alone, feel free to announce it in the slack group chat.
-mark it as resolved and remerge the PR again, there shouldn't any problem with it.

Feel free to add more about that matter here.

**Comment by NicoHambauer** [comment]
Hey there @ultiwinter ! Looks good, what about pushing deliverables files to the main? Do we also follow the guideline by creating a deliverable-n branch first where n is the respective sprint?

**Comment by Tims777** [comment]
I think that deliverables are probably the only thing where it does not make sense, to push them to dev first, because they are basically the result of the main branch (to be honest, they should not be in version control at all, but there's not a lot we can do about that)

**Comment by Tims777** [comment]
Unless we will be able to set up branch protection (which we currently do not have permission for), because then it would simply not be possible to push deliverables directly to main anymore

**Comment by felix-zailskas** [comment]
I think it might be good to not push to main to avoid main being ahead of dev. This merging strategy is also anchored in our team contract. This will force us to merge back from main at some point which is bad practice if not applying a hotfix. If our code is not ready and we need to push deliverables to main I would rather cherry pick the Deliverables folder to merge only the needed files into main.

**Comment by luccalb** [comment]
I believe uploading deliverables directly to main shouldn't be an issue because it's isolated in the deliverables folder right? When we then merge our code to main, git will realize theres no conflict and auto merge the changes afaik. It's true that it violates our branching structure, but it was meant to keep our codebase stable, and as theres never any code in the deliverables folder I don't see an issue.

I would like to add one more requirement before merging a PR: All GitHub Action checks should pass. Since the pipeline is setup, it will run for all branches/PRs (you might need to merge `dev` into your feature branch once).

**Comment by ultiwinter** [comment]
Having new deliverables pushed to main is not a problem. However, I believe
we would be better off having a standardized way and adhering to our
branching hierarchy. When we make new deliverables, merging into dev and
then directly merging dev into main shouldn’t be problem nor time
extensive. Hence, I add my voice to Felix’s opinion.

Lucca Baumgärtner ***@***.***> schrieb am So. 29. Okt. 2023
um 10:36:

**Comment by NicoHambauer** [comment]
Rules for who is among the Reviewers
